### PR TITLE
GRAM&ANN&TY: Support `if let guard` syntax

### PR DESCRIPTION
**Pull request by mchernyavsky**
Closes https://github.com/intellij-rust/intellij-rust/issues/7794.

changelog: Support [if let guard](https://rust-lang.github.io/rfcs/2294-if-let-guard.html) syntax

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/core/parser/RustParserDefinition.kt [hunk 1]
```diff
@@ -104,6 +104,6 @@ class RustParserDefinition : ParserDefinition {
         /**
          * Should be increased after any change of parser rules
          */
-        const val PARSER_VERSION: Int = LEXER_VERSION + 33
+        const val PARSER_VERSION: Int = LEXER_VERSION + 34
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/core/parser/RustParserDefinition.kt:107]:
```suggestion
        const val PARSER_VERSION: Int = LEXER_VERSION + 35
```

`34` version already exists

**Review by Undin — COMMENTED** [review]
It would be great to support type inference and name resolution for items from `if let` guard.
For example, in the following example
```rust
#![feature(if_let_guard)]

enum Foo {
    Bar(i32),
    Baz
}

fn foo(f: Foo, y: i32) {
    match y {
        1 if let Foo::Bar(z) = f => { z + y }
        _ => 0
    };
}
```

the plugin knows neither the type nor location of `z` variable

Don't insist on implementing it in this PR. But probably it's worth converting the original issue into a meta issue for `if let guard` feature support and track progress there

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt [hunk 2]
```diff
@@ -689,6 +690,17 @@ class RsErrorAnnotator : AnnotatorBase(), HighlightRangeExtension {
...
     }
 
+    private fun checkMatchArmGuard(holder: RsAnnotationHolder, guard: RsMatchArmGuard) {
+        if (guard.let != null) {
+            IF_LET_GUARD.check(holder, guard, "if let guard")
+        }
+
+        val pat = guard.pat
+        if (pat != null && pat.isIrrefutable) {
+            IRREFUTABLE_LET_PATTERNS.check(holder, pat, "irrefutable let pattern")
```
[the remark this answers is not in the document]
Comment (mchernyavsky) [reply src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt:700]: So, should I remove it or convert it into a lint inspection?

[the remark this answers is not in the document]
Comment (Undin) [reply src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt:700]: Drop from this PR
And if you would like - implement lint inspection in separate PR

**Comment by mchernyavsky** [comment]
@Undin fixed.

**Review by ozkriff — APPROVED** [review]

**Review by Undin — APPROVED** [review]
👍 
bors r+

**Pull request merged**